src/gdb-stub/gdb_stub.cpp

```diff
@@ -0,0 +1,492 @@
+            size_t kind{};
+            rt_assert(sscanf_s(data.c_str(), "%x,%" PRIX64 ",%zx", &type, &addr, &kind) == 3);
+
+            const auto res = change_breakpoint(handler, set, translate_breakpoint_type(type), addr, kind);
+            connection.send_reply(res ? "OK" : "E01");
+        }
+
+        void handle_v_packet(const connection_handler& connection, const std::string_view data)
+        {
+            const auto [name, args] = split_string(data, ':');
+
+            if (name == "Cont?")
+            {
+                // IDA pro gets confused if the reply arrives too early :(
+                std::this_thread::sleep_for(1s);
+
+                connection.send_reply("vCont;s;c;");
+            }
+            else
+            {
+                connection.send_reply({});
+            }
+        }
+
+        void read_registers(const connection_handler& connection, debugging_handler& handler)
+        {
+            std::string response{};
+            std::vector<std::byte> data{};
+            data.resize(handler.get_max_register_size());
+
+            const auto registers = handler.get_register_count();
+
+            for (size_t i = 0; i < registers; ++i)
+            {
+                memset(data.data(), 0, data.size());
+                const auto res = handler.read_register(i, data.data(), data.size());
+
+                if (!res)
+                {
+                    connection.send_reply("E01");
+                    return;
+                }
+
+                response.append(utils::string::to_hex_string(data));
+            }
+
+            connection.send_reply(response);
+        }
+
+        void write_registers(const connection_handler& connection, debugging_handler& handler,
+                             const std::string_view payload)
+        {
+            const auto data = utils::string::from_hex_string(payload);
+
+            const auto registers = handler.get_register_count();
+            const auto register_size = handler.get_max_register_size();
+
+            for (size_t i = 0; i < registers; ++i)
+            {
+                const auto offset = i * register_size;
+                const auto end_offset = offset + register_size;
+
+                if (data.size() < end_offset)
+                {
+                    connection.send_reply("E01");
+                    return;
+                }
+
+                const auto res = handler.write_register(i, data.data() + offset, register_size);
+
+                if (!res)
+                {
+                    connection.send_reply("E01");
+                    return;
+                }
+            }
+
+            connection.send_reply("OK");
+        }
+
+        void read_single_register(const connection_handler& connection, debugging_handler& handler,
+                                  const std::string& payload)
+        {
+            size_t reg{};
+            rt_assert(sscanf_s(payload.c_str(), "%zx", &reg) == 1);
```

Comment:
this reads the incorrect register, it should be %zd i believe

Reply:
Interesting. I copied the implementation from here: https://github.com/momo5502/mini-gdbstub/blob/main/lib/gdbstub.c#L196

Reply:
https://sourceware.org/gdb/current/onlinedocs/gdb.html/Packets.html#Packets

![image](https://github.com/user-attachments/assets/dd1eb528-4bfa-40f9-a79b-1c207eb4ae38)

%zx seems correct